src/treebuilders/apply.cpp

```diff
@@ -91,6 +91,64 @@ template <int D> void apply(double prec, FunctionTree<D> &out, ConvolutionOperat
     print::separator(10, ' ');
 }
 
+
+/** @brief Application of MW integral convolution operator on Four component
+ *
+ * @param[in] prec: Build precision of output function
+ * @param[out] out: Output function to be built
+ * @param[in] oper: Convolution operator to apply
+ * @param[in] inp: Input function
+ * @param[in] metric: 4x4 array with coefficients that relates the in and out components
+ * @param[in] maxIter: Maximum number of refinement iterations in output tree, default -1
+ * @param[in] absPrec: Build output tree based on absolute precision, default false
+ *
+ * @details The output function will be computed using the general algorithm:
+ * - For each input component apply the operator
+ * - Compute MW coefs on current grid
+ * - Refine grid where necessary based on `prec`
+ * - Repeat until convergence or `maxIter` is reached
+ * - `prec < 0` or `maxIter = 0` means NO refinement
+ * - `maxIter < 0` means no bound
+ * - After application multiply by metric coefficient, and put in relevant output component
+ *
+ * @note This algorithm will start at whatever grid is present in the `out`
+ * tree when the function is called (this grid should however be EMPTY, e.i.
+ * no coefs).
+ *
+ */
+template <int D> void apply(double prec, CompFunction<D> &out, ConvolutionOperator<D> &oper, const CompFunction<D> &inp, ComplexDouble metric[4][4], int maxIter, bool absPrec) {
+
+    ComplexDouble defaultMetric[4][4];
+    for (int i=0; i<4; i++){
+        for (int j=0; j<4; j++){
+            if (i==j) defaultMetric[i][j] = 1.0;
+            else defaultMetric[i][j] = 0.0;
+        }
+    }
+    if (metric == nullptr) {
```

Comment:
C++17 introduced [`std::optional`](https://en.cppreference.com/w/cpp/utility/optional) to deal with arguments that may or may not be passed. We should use that here.